tools/rostopic/src/rostopic/__init__.py

```diff
@@ -285,45 +285,59 @@ def _rostopic_bw(topic, window_size=-1):
         _sleep(1.0)
         rt.print_bw()
 
-# TODO: port to the version I wrote for rxplot instead as it should be more efficient
-
+# code adapted from rqt_plot
 def msgevalgen(pattern):
     """
-    Generates a function that returns the relevant field (aka 'subtopic') of a Message object
-    :param pattern: subtopic, e.g. /x. Must have a leading '/' if specified, ``str``
+    Generates a function that returns the relevant field(s) (aka 'subtopic(s)') of a Message object
+    :param pattern: subtopic, e.g. /x[2:]/y[:-1]/z, ``str``
     :returns: function that converts a message into the desired value, ``fn(Message) -> value``
     """
-    if not pattern or pattern == '/':
-        return None
-    assert pattern[0] == '/'
-    msg_attribute = pattern[1:]
-
-    # use slice arguments if present
-    array_index_or_slice_object = None
-    index = msg_attribute.find('[')
-    if index != -1:
-        if not msg_attribute.endswith(']'):
-            sys.stderr.write("Topic name '%s' contains '[' but does not end with ']'\n" % msg_attribute)
-            return None
-        index_string = msg_attribute[index + 1:-1]
-        try:
-            array_index_or_slice_object = _get_array_index_or_slice_object(index_string)
-        except AssertionError as e:
-            sys.stderr.write("Topic name '%s' contains invalid slice argument '%s': %s\n" % (msg_attribute, index_string, str(e)))
-            return None
-        msg_attribute = msg_attribute[:index]
+    evals = [] # list of (field_name, slice_object) pairs
+    fields = [f for f in pattern.split('/') if f]
+    for f in fields:
+        if '[' in f:
+            field_name, rest = f.split('[')
+            rest = rest[:rest.find(']')]
+            try:
+                array_index_or_slice_object = _get_array_index_or_slice_object(rest)
+            except AssertionError as e:
+                sys.stderr.write("field '%s' has invalid slice argument '%s': %s\n"
+                                 % (field_name, rest, str(e)))
+                return None
+            evals.append((field_name, array_index_or_slice_object))
+        else:
+            evals.append((f, None))
+
+    def _msgeval(msg, evals):
+        if not evals:
+            return msg
+        for i, (field_name, slice_object) in enumerate(evals):
+            try: # access field first
+                msg = getattr(msg, field_name)
+            except AttributeError:
+                sys.stderr.write("no field named [%s]\n" % field_name)
```

Comment:
May be more context information would be helpful for users. E.g. show the full "path" to the unresolvable field.

Same at the next except block.